### ISSUE TITLE
There seems to sometimes be an issue with the DI system when using IServiceScopeFactory

### ISSUE THREAD

**Issue by Unskilledcrab** [koenbeuk/EntityFrameworkCore.Triggered#124]
I'm not exactly sure what the issue is but the DI system is throwing an error when I use the IServiceScopeFactory and make a call the the database that will end up running one the Trigger system. From the stacktrace provided is seems to me like it is an issue within the Trigger library

Image of the issue
![image](https://user-images.githubusercontent.com/12536115/135940843-0e9c02ba-8014-4d79-8cbd-59aea893ecdb.png)

The caller of the issue
![image](https://user-images.githubusercontent.com/12536115/135940960-6755e0c2-99bc-4bec-b05f-c02a9e041bd6.png)

How the IServiceScopeFactory is injected into the constructor
![image](https://user-images.githubusercontent.com/12536115/135940991-fcb792e4-bcf8-49aa-9004-d2befee2231c.png)

Stack Trace of the issue

```
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ThrowHelper.ThrowObjectDisposedException()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at EntityFrameworkCore.Triggered.Internal.HybridServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetServices(IServiceProvider provider, Type serviceType)
   at EntityFrameworkCore.Triggered.Internal.TriggerFactory.<Resolve>d__3.MoveNext()
   at System.Linq.Enumerable.<SelectIterator>d__174`2.MoveNext()
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source, Int32& length)
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__17.MoveNext()
   at System.Linq.Enumerable.SelectIPartitionIterator`2.MoveNext()
   at System.Linq.Enumerable.<CastIterator>d__64`1.MoveNext()
   at EntityFrameworkCore.Triggered.TriggerSession.<RaiseAfterSaveFailedStartingTriggers>d__25.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at EntityFrameworkCore.Triggered.Internal.TriggerSessionSaveChangesInterceptor.<SaveChangesFailedAsync>d__9.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.EntityFrameworkCore.DbContext.<SaveChangesAsync>d__56.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at OTS.Data.Extensions.DbContextExtensions.<SaveAsNoTrackingAsync>d__1.MoveNext() in C:\Users\jb090\source\repos\OTS\OTS.Data\Extensions\DbContextExtensions.cs:line 64
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at OTS.DataAccess.Abstractions.BaseEntityService`1.<Update>d__8.MoveNext() in C:\Users\jb090\source\repos\OTS\OTS.DataAccess\Abstractions\BaseEntityService.cs:line 84
```

**Comment by Unskilledcrab** [comment]
Update: I just tried updating the library to see if that would fix the issue, I am now running **2.4.0-beta.1**. I thought that maybe the extension method **SaveChangesWithoutTriggersAsync()** would not need to call into the DI system so could be a temporary work around. Unfortunately this has the same issue still.

![image](https://user-images.githubusercontent.com/12536115/135942380-597971de-0c96-4987-b20e-25ffdaeae39e.png)

**Comment by koenbeuk** [comment]
In your `TryUpdateTagValues` method, you are calling `using var scope = scopeFactory.CreateScope();`. This will dispose the same scope (and thereby any services resolved through that scope) when the method ends. Since you don't await subsequent calls to TagService and TagEventConfigurationService, these will then likely run and do stuff after the scope is already resolved. In this case it's causing issues with the Triggered library but it can evenly affect other services that rely on the DI container.

An easy way to verify this is to await or synchronously wait eg. `tagService.Update(...).GetAwaiter().GetResult()`.

**Comment by Unskilledcrab** [comment]
Ah.. thats interesting. The work around that i am currently doing is Instead of calling EF Core save changes. I have made a separate method in my service that doesn't call the savechanges method in EF Core and therefore won't get intercepted by the triggered library

![Screenshot_20211005-081802_GitHub.jpg](https://user-images.githubusercontent.com/12536115/136020924-ead8da58-eb41-421f-a46a-e7c4062650dc.jpg)

I'm not awaiting the service update method call because it is in a timed background service and is something that I want to fire and forget and don't care about the result.

So what you are saying is that EF core doesn't rely on any of the DI service collection internally but your library does?

**Comment by koenbeuk** [comment]
@Unskilledcrab That is correct. Given that Triggers often interacts with application services, this library is capturing the ServiceProvider that was used to obtain the DbContext to resolve those services from:

```csharp
new ServiceCollection()
  .AddScoped<MyService>()
  .AddDbContext<MyDbContext>(options => options.UseTriggers(options => options.AddTriger<FooTrigger>())
  .AddTransiant<IBeforeSaveTrigger<Bar>, BarTrigger>()
``` 
In this example, FooTrigger may depend on MyService which gets provided by the ServiceProvider relevant to the caller of SaveChanges (which is almost always what we want). In addition, it is possible (as demonstrated with the BarTrigger) to register triggers directly with your DI container. These triggers will also be recognized and work equally well. 

In general, I would consider it as bad practice to dispose of the ServiceScope before finishing all related work (in this case with the DbContext). Imagine switching to a pooled strategy and bad things will happen :)

If you want to implement a fire and forget strategy then consider starting a new Task eg. `Task.Run(async () => { .... })`

**Comment by Unskilledcrab** [comment]
Ah 👍 👌  thank you so much for being so through with your answer, I learned quite a bit about some topics that have been difficult for me to understand. I've just recently started getting into multi threading and it seems like async and await cover up so much of what is going on behind the scenes.

So you are saying start the new task and within that new task, have the scoped service created in there and then await all calls using that scoped provider so that that scope isn't disposed of before all the calls using it are completed, and that would be the easiest solution to implement if I'm understanding you correctly. I will try to implement this solution and it seems like I have a bit of cleanup to do in my code base if that is best practices 😜 

Thanks again for the help! I suppose the issue should be closed since it doesn't follow best practices.
